File: main_project/tripod-new-branch/src/cpp/client/TripodClient.cpp
#include <TripodData.h>
#include <TripodClient.h>
#include <ClientFactory.h>
#include <CacheClient.h>
#include <RedisCacheClient.h>
#include <CacheManagerClient.h>

namespace com {
namespace renren {
namespace tripod {

TripodClient::TripodClient(const std::string& zkAddress, const std::string& namespaceId, 
                        const std::string& businessId) : 
                        zkAddress_(zkAddress), namespaceId_(namespaceId), businessId_(businessId) {
}

CacheClientPtr TripodClient::getCacheClient() {
    if(cacheClient_ == NULL) {
        boost::upgrade_lock<boost::shared_mutex> lock(cacheClientMutex_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        if(cacheClient_ == NULL) {
            cacheClient_ = ClientFactory::getInstance()->getCacheClient(zkAddress_);
        }
    }
    return cacheClient_;
}

CacheManagerClientPtr TripodClient::getCacheManagerClient() {
    if(cacheManagerClient_ == NULL) {
        boost::upgrade_lock<boost::shared_mutex> lock(cacheManagerClientMutex_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        if(cacheManagerClient_ == NULL) {
            cacheManagerClient_ = ClientFactory::getInstance()->getCacheManagerClient(zkAddress_);
        }
    }
    return cacheManagerClient_;
}

DataMap TripodClient::get(const KeySeq& keys, KeySeq& missedKeys, const long timeout) {
    CacheClientPtr client = getCacheClient();
    if(client == NULL) {
        return DataMap();
    }

    DataMap res = client->get(keys, namespaceId_, businessId_, timeout);
    
    getMissedKeys(res, keys, missedKeys);

    return res;
}

DataMap TripodClient::getMissed(const KeySeq& keys, long expiredTime, bool useMaster, const long timeout) {
    CacheManagerClientPtr client = getCacheManagerClient();
    if(client == NULL) {
        return DataMap();
    }

    return client->get(keys, namespaceId_, businessId_, expiredTime, useMaster, timeout);
}

DataMap TripodClient::getWithMissed(const KeySeq& keys, long expiredTime, bool useMaster, const long timeout) {
    KeySeq missedKeys;
    DataMap res = get(keys, missedKeys, timeout);

    if(!missedKeys.empty()) {
        DataMap missedRes = getMissed(missedKeys, expiredTime, useMaster, timeout);
        res.insert(missedRes.begin(), missedRes.end());
    }
    return res;
}


bool TripodClient::load(const KeySeq& keys, long expiredTime, bool useMaster, const long timeout) {
    CacheManagerClientPtr client = getCacheManagerClient();
    if(client == NULL) {
        return false;
    }

    return client->load(keys, namespaceId_, businessId_, expiredTime, useMaster, timeout);
}

bool TripodClient::deleteCache(const KeySeq& keys, const long timeout) {
    CacheManagerClientPtr client = getCacheManagerClient();
    if(client == NULL) {
        return false;
    }
    
    return client->deleteCache(keys, namespaceId_, businessId_, timeout);
}

void TripodClient::getMissedKeys(const DataMap& res, const KeySeq& keys, KeySeq& missedKeys) {
    for(KeySeq::const_iterator it = keys.begin(); it != keys.end(); ++it) {
        if(res.find(*it) == res.end()) {
            missedKeys.push_back(*it);
        }
    }
}



RedisCacheClientPtr TripodClient::GetRedisCacheClient() {
  if(cacheClient_ == NULL) {
    boost::upgrade_lock<boost::shared_mutex> lock(redis_cache_client_mutex_);
    boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
    if(redis_cache_client_ == NULL) {
       MCE_DEBUG("zkAddress_" << zkAddress_ );
       redis_cache_client_ = ClientFactory::getInstance()->GetRedisCacheClient(zkAddress_);
    }
  }
  return redis_cache_client_;
}

void TripodClient::GetList(const std::string& key, StrList& value, int begin, int limit) {
  RedisCacheClientPtr client = GetRedisCacheClient();
  if (!client) {
    MCE_INFO("TripodClient::GetList() RedisCacheClient is NULL");
    return;
  }
  const StrList& list = client->GetList(key, begin, limit, namespaceId_, businessId_);
  if (!list.empty()) {
    value.insert(value.end(), list.begin(), list.end());
  }
}


void TripodClient::GetIntList(const std::string& key, IntList& value, int begin, int limit) {
  RedisCacheClientPtr client = GetRedisCacheClient();
  if (!client) {
    MCE_INFO("TripodClient::GetIntList() RedisCacheClient is NULL");
    return;
  }
  const IntList& list = client->GetIntList(key, begin, limit, namespaceId_, businessId_);
  if (!list.empty()) {
    value.insert(value.end(), list.begin(), list.end());
  }
}

bool TripodClient::LoadListCache(const std::string& key, bool use_master) {
  CacheManagerClientPtr client = getCacheManagerClient();
  if(client == NULL) {
      return false;
  }
  client->LoadListCache(key, namespaceId_, businessId_, use_master);
  return true;
  
}

bool TripodClient::DeleteListCache(const std::string& key) {
  CacheManagerClientPtr client = getCacheManagerClient();
  if(client == NULL) {
      return false;
  }
  client->DeleteListCache(key, namespaceId_, businessId_);
  return true;
}

}}}

File: main_project/tripod-new-branch/src/java/TripodClient/src/main/java/com/renren/tripod/client/TripodClient.java
package com.renren.tripod.client;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.renren.tripod.internalclient.CacheClient;
import com.renren.tripod.internalclient.CacheManagerClient;
import com.renren.tripod.internalclient.ClientFactory;
import com.xiaonei.xce.log.Tripod;

public class TripodClient {
	
	private String _zkAddress = null;
	private String _namespaceId = null;
	private String _businessId = null;

	private CacheClient _cacheClient = null;
	private CacheManagerClient _cacheManagerClient = null;

	public TripodClient(String zkAddress, String namespaceId, String businessId) {
		_zkAddress = zkAddress;
		_namespaceId = namespaceId;
		_businessId = businessId;
	}

	public Map<String, String> get(List<String> keys, List<String> missedKeys) {
		CacheClient cc = getCacheClient();
		if (cc == null) {
			Tripod.getLogger().error("TripodClient.get cacheclient == null for zkAddress "
					+ _zkAddress);
			return null;
		}
		Map<String, String> res = cc.get(keys, _namespaceId, _businessId);
		getMissedKeys(res, keys, missedKeys);
		return res;
	}

	public Map<String, String> getMissed(List<String> keys, long expiredTime,
			boolean useMaster) {
		return getMissed(keys, expiredTime, useMaster, 300);
	}

	public Map<String, String> getMissed(List<String> keys, long expiredTime,
			boolean useMaster, int timeout) {
		CacheManagerClient cc = getCacheManagerClient();
		if (cc == null) {
			Tripod.getLogger().error("TripodClient.getMissed cachemanagerclient == null for zkAddress "
					+ _zkAddress);
			return null;
		}
		return cc.get(keys, _namespaceId, _businessId, expiredTime, useMaster,
				timeout);
	}
	
	public Map<String, String> getWithMissed(List<String> keys, long expiredTime,
			boolean useMaster, int timeout) {
		List<String> missedKeys = new ArrayList<String>();
		Map<String, String> res = get(keys, missedKeys);
		
		if(!missedKeys.isEmpty()) {
			Map<String, String> missedRes = getMissed(missedKeys, expiredTime, useMaster, timeout);
			res.putAll(missedRes);
		}
		return res;
	}

	public boolean load(List<String> keys, long expiredTime, boolean useMaster) {
		return load(keys, expiredTime, useMaster, 1000);
	}

	public boolean load(List<String> keys, long expiredTime, boolean useMaster,
			int timeout) {
		CacheManagerClient cc = getCacheManagerClient();
		if (cc == null) {
			Tripod.getLogger().error("TripodClient.load cachemanagerclient == null for zkAddress "
					+ _zkAddress);
			return false;
		}
		return cc.load(keys, _namespaceId, _businessId, expiredTime, useMaster,
				timeout);
	}

	public boolean deleteCache(List<String> keys) {
		return deleteCache(keys, 1000);
	}

	public boolean deleteCache(List<String> keys, int timeout) {
		CacheManagerClient cc = getCacheManagerClient();
		if (cc == null) {
			Tripod.getLogger().error("TripodClient.deleteCache cachemanagerclient == null for zkAddress "
					+ _zkAddress);
			return false;
		}
		return cc.deleteCache(keys, _namespaceId, _businessId, timeout);
	}

	private List<String> getMissedKeys(Map<String, String> res,
			List<String> keys, List<String> missedKeys) {
		for (String str : keys) {
			if (res.get(str) == null) {
				missedKeys.add(str);
			}
		}
		return missedKeys;
	}

	private CacheClient getCacheClient() {
		if (_cacheClient == null) {
			synchronized (this) {
				if (_cacheClient == null) {
					_cacheClient = ClientFactory.getCachegClient(_zkAddress);
				}
			}
		}
		return _cacheClient;
	}

	private CacheManagerClient getCacheManagerClient() {
		if (_cacheManagerClient == null) {
			synchronized (this) {
				if (_cacheManagerClient == null) {
					_cacheManagerClient = ClientFactory
							.getCacheManagerClient(_zkAddress);
				}
			}
		}
		return _cacheManagerClient;
	}
	
	public static void main(String[] args) {
		
		//PropertyConfigurator.configure("/Volumes/Work/code/forxcecache/tripod-new/src/java/TripodClient/src/main/java/com/renren/tripod/client/log4j.properties");
		TripodClient tc = new TripodClient("TestZooKeeper1:2181/Tripod", "test", "TEST");
		List<String> keys = new ArrayList<String>();
		keys.add("123");
		
		System.out.println("load " + tc.load(keys, 1000, false));
		
		List<String> missedKeys = new ArrayList<String>();
		Map<String, String> res = tc.get(keys, missedKeys);
		System.out.println("get size " + res.size());
		System.out.println("get size " + res.get("123"));
		System.out.println("get missed size " + missedKeys.size());
		
		System.out.println("delete " + tc.deleteCache(missedKeys));
		
		res = tc.getMissed(keys, 1000, false);
		System.out.println("getmissed size " + res.size());
		
		System.exit(0);
		
	}
	
}
